File: src/cpp/explore/topquestions/easy/array/旋转数组.cpp
/**
 * @File    :   旋转数组.cpp
 * @Time    :   2020/07/24 07:46:19
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :   https://leetcode-cn.com/explore/featured/card/top-interview-questions-easy/1/array/23/
 */

#include <vector>
using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        if (k >= n) {
            k %= n;
        }

        rotate(nums, 0, n - k - 1);
        rotate(nums, n - k, n - 1);
        rotate(nums, 0, n - 1);
    }

    void rotate(vector<int>& nums, int left, int right) {
        for (int i = left, j = right; i < j; i++, j--) {
            swap(nums[i], nums[j]);
        }
    }
};

File: src/python/explore/topquestions/easy/array/旋转数组.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   旋转数组.py
@Time    :   2020/07/26 09:28:12
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        n = len(nums)
        if n == 0:
            return
        k %= n

        self.reverse(nums, 0, n - k - 1)
        self.reverse(nums, n - k, n - 1)
        self.reverse(nums, 0, n - 1)

    def reverse(self, nums, i, j):
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1
